pkg/rate/api_limiter_test.go

```diff
@@ -716,3 +718,78 @@ func (b *ControllerSuite) TestAdjustedParallelRequests(c *check.C) {
 	a.adjustmentFactor = 1.0
 	c.Assert(a.adjustedParallelRequests(), check.Equals, 2)
 }
+
+func (b *ControllerSuite) TestReservationCancel(c *check.C) {
+	a := NewAPILimiter("foo", APILimiterParameters{
+		RateLimit:           50.0,
+		RateBurst:           10,
+		ParallelRequests:    1,
+		MaxParallelRequests: 1,
+		MaxWaitDuration:     500 * time.Millisecond,
+		Log:                 true,
+	}, nil)
+
+	// Process a request but don't complete it, this will occupy the
+	// parallel request slot
+	req, err := a.Wait(context.Background())
+	c.Assert(err, check.IsNil)
+
+	var completed int32
+
+	// All of these requests must fail due to having to wait too long as
+	// the only parallel request slot is occupied. The rate limiter should
+	// not get occupied with these requests though.
```

Comment:
> The rate limiter should not get occupied with these requests though.

Did you intend to test this as well?

This test passes on master, regardless of the patches. Even the error returned by each goroutine is the same with/without this PR:

```
timed out while waiting to be served with 1 parallel requests: context deadline exceeded
```

The test itself looks like a reasonable assessment of the behaviour though, just checking whether it was intended as a general improvement in coverage or whether you were aiming to test the functionality of this PR specifically.

Reply:
I've added the test to validate the cancelation of the parallel request and did not find any bugs in this regard in the existing code. I figured it makes sense to have the test in the codebase though to catch eventual regressions.